### ISSUE TITLE
Test suite for HTTP Gateways

### ISSUE THREAD

**Issue by lidel** [ipfs/in-web-browsers#205]
## Current state

- Specifications for HTTP Gateways exist now at [ipfs/specs/http-gateways](https://github.com/ipfs/specs/blob/main/http-gateways)
- For historical reasons, comprehensive, ent-to-end HTTP Gateway tests live in Kubo IPFS implementation (`*gateway*.sh` in [kubo/test/sharness](https://github.com/ipfs/go-ipfs/tree/master/test/sharness))
- We have only four simple tests at https://ipfs.github.io/public-gateway-checker/, and they are not robust enough.

## Target audience

- Gateway operators wanting to ensure their reverse proxy/cdn setup is correct
- Gateway implementer, ensuring they did not miss any nuance / edge case around [ipfs/specs/http-gateways](https://github.com/ipfs/specs/blob/main/http-gateways)
- IPFS user making decision which gateway is useful for their needs

## Needs

- Ability to run tests against HTTP endpoint and get a compliance report against [ipfs/specs/http-gateways](https://github.com/ipfs/specs/blob/main/http-gateways)
- Being able to test each gateway type separately  (e.g., Block/CAR gateway without anything else)
- Runs via CLI and on CI
- Human readable errors with references to relevant section of  [ipfs/specs/http-gateways](https://github.com/ipfs/specs/blob/main/http-gateways)
- (nice to haves)
  - Ability to generate HTML reports
  - Ability to reuse on https://ipfs.github.io/public-gateway-checker/ (pre-generate tests on CI for things that can't be tested live via JS)

**Comment by lidel** [comment]
The caveat is that there are HTTP tests which can't be done via JS running on a webpage (due to things like access to HTTP headers, redirects, CORS etc). We will never have a truly isomorphic test suite. Checker would be consuming this, but we should not design it around that use case.  The target audience is different.

@darobin may have some ideas how to get the biggest bang for the buck here

**Comment by darobin** [comment]
I would also worry that the test suite could become rather large as more corner cases get covered, which might not align well with what is good for `public-gateway-checker` (I'm not sure it's supposed to be fully comprehensive).

What I have in mind is (roughly, with ideas taken from WPT):
* The primary use case is CI. This isn't to say that other uses are precluded, just that the goal is interop and what's in CI gets fixed. This provides an incentive for users who find an interop issue to contribute a test case.
* The test runner gets passed the location of the gateway (it could also start the daemon itself, WPT does that with Web Driver, we can look into that).
* Reporting in various formats is easy. I'm not sure what you have in mind with "previews on PRs" @lidel? It would also be easy to report in MD, meaning the output could then just be pasted into the PR description if that's useful.
* This should "just" be a subset of the full test suite, but running just a subset (or even a single test) ought to be easy.
* "Human readable errors" likely depends more on the tests than on the test runner. However, I was considering the idea that we could have the tests and the specs in the same repo (basically the interop monorepo). So the tests in `tests/ipfs/gateways/*` would be for the spec in `specs/ipfs/gateways.md`. That means the tests could easily reference notions defined in the specs (because the metadata for that is automatically exported when the spec is generated) so you could eg. have a test string stating `check that the :ref[OPTIONS method] lists all verbs` and have the report link to the right bit of the spec.
* Conversely, running the test with a `--save` flag would output results for that implementation to the same repo, which in turn the specs could use to showcase how well supported they are, à la caniuse.com.

Anyway, these are just a few thoughts, I'm not married to anything — this needs to work for implementers, not me! As I was saying in Slack, I'm happy to bang something together so folks can see if they like/hate it.

**Comment by darobin** [comment]
Ok, I made a prototype — let me know if you think this is headed in the right direction. The runner is over at https://github.com/darobin/ipseity/blob/main/bin/bastest.js, and I made a quick and dirty port of the Kubo block tests at https://github.com/darobin/ipseity/blob/main/test-suite/ipfs/gateways/blocks.js.

The runner expects a gateway, and can be told to be quiet, to save JSON, and to output Markdown (it can do these in parallel).

A failing gateway: `./bin/bastest.js --gateway https://berjon.com/ `:

<img width="964" alt="image" src="https://user-images.githubusercontent.com/38491/203179220-f6c63783-1269-496b-a1e6-61ab3fcc0631.png">

A successful gateway (assuming `ipfs daemon --writable`) run that also saves JSON and MD: `./bin/bastest.js --gateway http://127.0.0.1:8080/ --save robin-test_0.0.1 --markdown scratch/test.md`:

<img width="964" alt="image" src="https://user-images.githubusercontent.com/38491/203179465-d189876e-2300-4a84-b41c-3ac20c4743a0.png">

The JSON it saves looks like this: https://github.com/darobin/ipseity/blob/main/test-reports/robin-test_0.0.1.json. Nothing phenomenal (and we'd have to figure out some useful conventions to produce interop reporting from it) but we can figure out workflows/configurations.

The MD it outputs is like this, just pasting it raw here:

# Interplanetary Test Suite
## Test HTTP Gateway Raw Block (application/vnd.ipld.raw) Support
### GET unixfs dir root block and compare it with expected raw block
* ✅ GET with format=raw param returns a raw block
* ✅ GET for application/vnd.ipld.raw returns a raw block
### Make sure expected HTTP headers are returned with the block bytes
* ✅ GET response for application/vnd.ipld.raw has expected response headers
* ✅ GET for application/vnd.ipld.raw with query filename includes Content-Disposition with custom filename
* ✅ GET response for application/vnd.ipld.raw has expected caching headers

---

Anyway — thoughts welcome!

**Comment by darobin** [comment]
Oh — I forgot an important point: one limitation of this approach is that the gateway must be at least minimally writeable. Otherwise there's no way to put the fixtures there. There could be some other ways of doing the same, but I don't think that they are any better. Unless there's a better idea that I haven't considered, I suspect that this is the best option.

**Comment by lidel** [comment]
I think this type of runner would do the trick, at least for gateways.
Rewriting all the tests will be tedious, but doable in chunks. 

@darobin for I/O-less tests we can leverage inlined CIDs  (use `indentity` multicodec instead of hash digest).

```
$ echo 'hello inlined' | ipfs add --raw-leaves --inline 
added bafkqadtimvwgy3zanfxgy2lomvsau bafkqadtimvwgy3zanfxgy2lomvsau
```

CID [bafkqadtimvwgy3zanfxgy2lomvsau](https://cid.ipfs.tech/#bafkqadtimvwgy3zanfxgy2lomvsau)  includes data in the multihash  digest, which means resolving it does not require IO/networking,  it will work in offline mode without networking too too.

**Comment by galargh** [comment]
#### Existing tests

I created a high-level overview of tests we currently have living in Kubo - https://gist.github.com/galargh/74f2ecaf9da2f590756785694719566c

Findings (that are probably obvious for all involved here already but helped me reason about the space):
- most of the tests rely on Kubo daemon for creating entities (`ipfs add`)
  - Could we decouple entity creation from the actual test? We could prepare a set of entities that must be available on the gateway for the tests and a script that uses a writeable gateway API to publish these entities. If a gateway supports the write API, it could use the script. Otherwise, the gateway owner would have to take care of entity creation.
- some of the tests rely on the Kubo daemon for verification of gateway responses (`$RESPONSE == $(ipfs cat)`)
  - Would it be possible always to use concrete expected values?
- some of the tests reconfigure gateway settings during the run
  - Could we split such tests so that all the test cases within a single test would be expected to pass with an immutable gateway configuration?
- some of the tests spoof the `Origin` header
- other than that, the tests follow a pretty standard pattern of request and verify response code, headers and body

If we could separate Gateway configuration, including entity creation, from tests, could the tests become a collection of fixed request and response parameters?

Finally, having these tests provides a fantastic opportunity. We can leverage it to ensure the new test suite works as expected by running them in tandem. 

#### Where should this live?

> I was considering the idea that we could have the tests and the specs in the same repo

I like this idea. It would enable a process where spec changes would have to be backed by accompanying tests. I wonder if we could somehow leverage a tool like https://github.com/warpfork/go-testmark to tie the two more closely together.

We should also make it easy for specs contributors to contribute to the testing suite. What language are they most likely familiar with? Are there any testing frameworks they are already more familiar with that we could lean on?

#### Reporting

> The runner is over at https://github.com/darobin/ipseity/blob/main/bin/bastest.js

Would it make sense to reuse existing reporters?

**Comment by darobin** [comment]
A few quick notes:
* Yes, we could ensure that a number of fixtures are hosted and use those. We can also use the `identity` multicodec for at least some of the tests.
* We can definitely use concrete expected values if we have them locally, which I reckon we should.
* We should assume an immutable gateway configuration; if there are different useful behaviours that could be exposed independently, maybe we should consider making them part of the protocol?
* Spoofing `Origin`: not a big deal, we can set arbitrary headers. Or am I missing the issue?
* Fixed request/response: some bits aren't going to be regular or consistent (notably some headers) so I don't think that it can be purely fixed. However, I think it's not far from the truth so it should be almost that.
* There has been a lot of hesitation around having everything in a monorepo so I don't think that'll fly but happy to chat about it.
* Reusing reporters: Mocha only supports one reporter at a time and that thing reports to three outputs at the same time. It's just a quick hack, we don't have to do it that way. One alternative is to report to TAP and convert. (It's not less code or necessarily simpler, though.)

**Comment by darobin** [comment]
Re `mocha-multi` I tried it, but it didn't like me at all so I gave up fighting it (the docs do say that it's a pile of hacks) and just put it all in there :) But if it works then 🚀!

I like your thing, it looks great and very usable! Really cool stuff.

I don't believe that there's a reliable way to create dir that way — I put it directly in the fixture.

**Comment by laurentsenta** [comment]
(note we're exploring options for this work in https://github.com/ipfs/gateway-conformance)
